### PR TITLE
Refactor builder to module

### PR DESCRIPTION
**Pull request by mgkwill**
Issue Number: #169

Objective of pull request: Refactor builder into module builders

## Pull request checklist

Your PR fulfills the following requirements:
- [x] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`pyb`) passes locally
- [x] Build tests (`pyb -E unit`) or (`python -m unittest`) passes locally

## Pull request type

Please check your PR type:
- [x] Refactoring (no functional changes, no api changes)

## What is the current behavior?

- builder.py is currently part of the src/lava/magma/compiler/ module

## What is the new behavior?

- builder.py is refactored, to an abstract interfactes.py and moved to module src/lava/magma/compiler/builders module

## Does this introduce a breaking change?

- [x] No

### CODE REVIEW

File: src/lava/magma/compiler/builders/interfaces.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+from abc import ABC, abstractmethod
+
+import typing as ty
+
+from lava.magma.compiler.channels.interfaces import AbstractCspPort
+if ty.TYPE_CHECKING:
+    from lava.magma.core.model.model import AbstractProcessModel
+
+
+class AbstractProcessBuilder(ABC):
```
Comment (awintel) [inline src/lava/magma/compiler/builders/interfaces.py]: Let's get in the habit of writing class docstrings right away. 
Here we should state what Builders are and how this abstract class is supposed to be used.

File: src/lava/magma/compiler/builders/interfaces.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    def set_csp_ports(self, csp_ports: ty.List[AbstractCspPort]):
+        pass
+
+    @property
+    @abstractmethod
+    def proc_model(self) -> "AbstractProcessModel":
+        pass
+
+
+class AbstractRuntimeServiceBuilder(ABC):
```
Comment (awintel) [inline src/lava/magma/compiler/builders/interfaces.py]: +docstring
+type annotations for __init__ arguments

File: src/lava/magma/compiler/builder/builder.py [hunk 3]
```diff
@@ -28,7 +32,8 @@
     AbstractPyPort,
     PyInPort,
     PyOutPort,
-    PyRefPort, PyVarPort,
+    PyRefPort,
+    PyVarPort
 )
 from lava.magma.compiler.channels.interfaces import AbstractCspPort, Channel, \
     ChannelType
```
Comment (awintel) [inline src/lava/magma/compiler/builder/builder.py:39]: See the next Todo comment. While you are at factoring out the AbstractProcBuilder class you could also more generic attributes like _proc_model, _model_id, _proc_params out to the abstract class.

File: src/lava/magma/compiler/compiler.py [hunk 4]
```diff
@@ -16,11 +16,11 @@
...
 import lava.magma.compiler.exceptions as ex
 import lava.magma.compiler.exec_var as exec_var
-from lava.magma.compiler.builders.builder import ChannelBuilderMp
-from lava.magma.compiler.builders.builder import PyProcessBuilder, \
+from lava.magma.compiler.builder.builder import ChannelBuilderMp
+from lava.magma.compiler.builder.builder import PyProcessBuilder, \
     AbstractRuntimeServiceBuilder, RuntimeServiceBuilder, \
     AbstractChannelBuilder, ServiceChannelBuilderMp
-from lava.magma.compiler.builders.builder import RuntimeChannelBuilderMp
+from lava.magma.compiler.builder.builder import RuntimeChannelBuilderMp
```
Comment (awintel) [inline src/lava/magma/compiler/compiler.py:23]: Wouldn't "builders" be more appropriate because this package contains multiple builders?

**Review by awintel — APPROVED** [review]
Looks good. I just added a few requests/comments here and there to add documentation.
If we touch undocumented code, it's always a good chance to add missing docstrings. 

Thanks

**Review by PhilippPlank — APPROVED** [review]
Looks good to me, beside the already requested docstrings.

File: src/lava/magma/compiler/compiler.py (hunk 4 above)
Comment (mgkwill) [reply src/lava/magma/compiler/compiler.py:23]: I used builders originally. I'll restore to that version.

I was trying to match some other module naming in lava, it seems we are inconsistent with naming.

File: src/lava/magma/compiler/builder/builder.py (hunk 3 above)
Comment (mgkwill) [reply src/lava/magma/compiler/builder/builder.py:39]: ACK

File: src/lava/magma/compiler/builders/interfaces.py (hunk 2 above)
Comment (mgkwill) [reply src/lava/magma/compiler/builders/interfaces.py]: ACK

**Review by GaboFGuerra — APPROVED** [review]
Looks good! Though it seems unit tests are failing with: "No module named 'lava.magma.compiler.builders'"

**Pull request merged**